docs/design.md

```diff
@@ -0,0 +1,382 @@
+no data is received, the connection should be presumed closed and a new
+connection started.
+
+A UserAgent MUST NOT send a Ping more frequently than once a minute or its
+connection MAY be dropped.
+
+The UserAgent should consider normal communications as an indication that the
+socket is working properly. It SHOULD send the ping packet only if no activity
+has occurred in the past 30 minutes.
+
+**Note**: This section is included for relevant historical purposes as the
+current implementation sends WebSocket Ping frames every 5 minutes which is
+sufficient to keep the connection alive. As such, client-sent Ping's are no
+longer needed.
+
+#### UserAgent -> PushServer
+
+The 2-character string {} is sent. This is a valid JSON object that requires no
+alternative processing on the server, while keeping transmission size small.
+
+#### PushServer -> UserAgent
+
+The PushServer may reply with any data. The UserAgent is only concerned about
+the state of the connection. The PushServer may deliver pending notifications or
+other information. If there is no pending information to be sent, it is
+RECOMMENDED that the PushServer also reply with the string {}.
+
+### Notification
+
+#### AppServer -> PushServer
+
+The AppServer MUST make a HTTP PUT request to the Endpoint received from the
+App, or a HTTP POST if using the WebPush extension.
+
+If no request body is present, the server MAY presume the version to be the
+current server UTC.
+
+If the request body is present, the request MUST contain the string "version=N"
+and the Content-Type MUST be application/x-www-form-urlencoded.
+
+#### PushServer -> AppServer
+
+The HTTP response status code indicates if the request was successful.
+
+* 200 - OK. The PushServer will attempt to deliver a notification to the
+  associated UserAgent.
+
+* 500 - Something went wrong with the server. Rare, but the AppServer should try
+  again.
+
+The HTTP response body SHOULD be empty.
+
+#### PushServer -> UserAgent
+
+Notifications are acknowledged by the UserAgent. PushServers should retry unacknowledged notifications every 60 seconds. If the version of an unacknowledged notification is updated, the PushServer MAY queue up a new notification for this channelID and the new version, and remove the old notification from the pending queue.
+
+messageType = "notification"
+
+updates list REQUIRED
+: The list contains one or more {"channelID": "id", "version": N } pairs.
+
+With the webpush extension to SimplePush, the following attributes are present
+in each notification instead of the single ``updates`` list above. A
+notification message is sent for every stored notification individually, as well
+as new notifications.
+
+channelID string REQUIRED
+: ChannelID of the notification
+
+version string REQUIRED
+: Version of the notification sent
+
+headers map OPTIONAL
+: Encryption headers sent along with the data. Present only if data was sent.
+
+data string OPTIONAL
```

Comment:
Indeed, was already merged though, so I'll file this for another bug. If you see anything else add it to #10.